core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -302,7 +325,11 @@ public void run() {
         listenerToClose.closed(reason, new Metadata());
       }
       drainPendingCalls();
+      if (!isStreamTransferCompleted()) {
+        realStreamStarted.countDown();
+      }
     }
+    awaitStreamTransferCompletion();
```

Comment:
Hm, it seems not necessarily. 
If the real stream is set, (L127) drainingPendingCall is not in sync path, so there is still some time window where the  `delegation` flag is set but the call not drained, i.e. start() not executed. So we await.
Also, in non delegation case, wait for noop count down should not hurt.

Reply:
But still, `cancel()` itself doesn't need to be waiting on the transfer to complete, right?

Reply:
Good catch, it seems can be remove here. Because normally cancel() should not be waiting for transfer to complete. And when the last item was cancelled, I already checked the shutdown status in the subclass and would call `awaitStreamTransferCompletion()` for termination.